File: cli-ts/src/internal/phases/RecPhase.ts
/**
 * TypeScript port of RecPhase abstract class
 * Represents a computation of a set of elements done in phases with dependency resolution
 */

import { none, type Option, some } from "fp-ts/Option";
import { PhaseCache } from "./PhaseCache";
import type { Phase } from "./types";

/**
 * Abstract base class for recursive phase definitions
 * Each phase can compute and express dependencies for the next phase
 */
export abstract class RecPhaseBase<Id, T> {
	/**
	 * Type aliases for easier access to generic types
	 */
	readonly _Id!: Id;
	readonly _T!: T;

	/**
	 * Creates a new phase that follows this one
	 */
	next<TT>(f: Phase<Id, T, TT>, name: string): RecPhaseBase<Id, TT> {
		return new RecPhaseNext(this, f, new PhaseCache<Id, TT>(), name);
	}

	/**
	 * Conditionally creates a new phase that follows this one
	 * If the phase is None, returns the current phase unchanged
	 */
	nextOpt(op: Option<Phase<Id, T, T>>, name: string): RecPhaseBase<Id, T> {
		return op._tag === "Some"
			? new RecPhaseNext(this, op.value, new PhaseCache<Id, T>(), name)
			: this;
	}
}

/**
 * Initial phase that simply passes through the input
 */
export class RecPhaseInitial<Id> extends RecPhaseBase<Id, Id> {
	readonly _tag = "Initial" as const;
}

/**
 * A phase that transforms the output of a previous phase
 */
export class RecPhaseNext<Id, T, TT> extends RecPhaseBase<Id, TT> {
	readonly _tag = "Next" as const;
	readonly _TT!: TT;

	constructor(
		readonly prev: RecPhaseBase<Id, T>,
		readonly trans: Phase<Id, T, TT>,
		readonly cache: PhaseCache<Id, TT>,
		readonly name: string,
	) {
		super();
	}
}

// Type alias for compatibility
export type RecPhase<Id, T> = RecPhaseBase<Id, T>;

/**
 * Factory functions for creating RecPhase instances
 */
export const RecPhase = {
	/**
	 * Creates an initial phase for the given Id type
	 */
	apply: <Id>(): RecPhase<Id, Id> => new RecPhaseInitial<Id>(),

	/**
	 * Type guard for Initial phase
	 */
	isInitial: <Id>(phase: RecPhase<Id, any>): phase is RecPhaseInitial<Id> => {
		return (phase as any)._tag === "Initial";
	},

	/**
	 * Type guard for Next phase
	 */
	isNext: <Id, T>(
		phase: RecPhase<Id, T>,
	): phase is RecPhaseNext<Id, any, T> => {
		return (phase as any)._tag === "Next";
	},
};

/**
 * Utility functions for working with RecPhase
 */

/**
 * Creates an optional phase transformation
 */
export const optionalPhase = <Id, T>(
	phase: Option<Phase<Id, T, T>>,
): Option<Phase<Id, T, T>> => phase;

/**
 * Helper to create Some phase
 */
export const somePhase = <Id, T>(
	phase: Phase<Id, T, T>,
): Option<Phase<Id, T, T>> => some(phase);

/**
 * Helper to create None phase
 */
export const nonePhase = <Id, T>(): Option<Phase<Id, T, T>> => none;

File: cli/src/org/scalablytyped/converter/internal/phases/RecPhase.scala
package org.scalablytyped.converter.internal
package phases

/** A representation of a computation of a set of elements which is done in phases. Each phase can both compute and
  * express a new set of dependencies needed for the next phase.
  *
  * Each element passing through a phase is guaranteed to be passed its dependencies after having passed through that
  * given phase.
  *
  * @tparam Id
  *   Every element and dependency must have an id
  * @tparam T
  *   The type of the element at the current stage of the computation
  */
sealed abstract class RecPhase[Id, T] {
  final type _Id = Id
  final type _T  = T

  /** Append another phase to the pipeline with the supplied display `name`. */
  final def next[TT](f: Phase[Id, T, TT], name: String): RecPhase[Id, TT] =
    RecPhase.Next(this, f, new PhaseCache, name)

  /** Optionally append a phase when the `Option` is defined; otherwise return the current pipeline unchanged. */
  final def nextOpt(op: Option[Phase[Id, T, T]], name: String): RecPhase[Id, T] =
    op match {
      case None    => this
      case Some(p) => RecPhase.Next(this, p, new PhaseCache, name)
    }
}

object RecPhase {
  /** Create a phase chain whose first stage simply echoes the supplied ids. */
  def apply[Id]: RecPhase[Id, Id] = Initial()

  /** Marker phase that represents the starting point of the pipeline. */
  final case class Initial[Id]() extends RecPhase[Id, Id]

  /**
    * A concrete transformation node that depends on `prev`, runs `trans`, and caches results under `name`.
    */
  final case class Next[Id, T, TT](
      prev: RecPhase[Id, T],
      trans: Phase[Id, T, TT],
      cache: PhaseCache[Id, TT],
      name: String
  ) extends RecPhase[Id, TT] {
    type _TT = TT
  }
}
